### PR TITLE
Add ARP app

### PR DESCRIPTION
**Pull request by wingo**
This PR pulls out the ARP app originally written by @kbara as part of the lwAFTR project to a more standalone, documented position.  It also changes it so that ARP will stamp outgoing ethernet traffic with the source L2 address.  This lets more Snabb programs be able to ignore link-layer details.

### CODE REVIEW

File: src/apps/ipv4/arp.lua [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+         -- Packet too short.
+         packet.free(p)
+      elseif is_arp(p) then
+         local h = ffi.cast(ether_arp_header_ptr_t, p.data)
+         if (ntohs(h.arp.htype) ~= arp_htype_ethernet or
+             ntohs(h.arp.ptype) ~= arp_ptype_ipv4 or
+             h.arp.hlen ~= 6 or h.arp.plen ~= 4) then
+            -- Ignore invalid packet.
+         elseif ntohs(h.arp.oper) == arp_oper_request then
+            if self.self_ip and ipv4_eq(h.arp.tpa, self.self_ip, 4) then
```
Comment (dpino) [inline src/apps/ipv4/arp.lua]: `ipv4_eq` always compares 4 bytes, so the third parameter in the function call can be removed.

File: src/apps/ipv4/arp.lua [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+               transmit(osouth, make_arp_reply(self.self_mac, self.self_ip,
+                                               h.arp.sha, h.arp.spa))
+            end
+         elseif ntohs(h.arp.oper) == arp_oper_reply then
+            if ipv4_eq(h.arp.spa, self.next_ip, 4) then
```
Comment (dpino) [inline src/apps/ipv4/arp.lua]: Likewise.

File: src/apps/ipv4/arp.lua [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+end
+
+local function is_arp(p)
+   if p.length < ether_arp_header_len then return false end
+   local h = ffi.cast(ether_arp_header_ptr_t, p.data)
+   return ntohs(h.ether.type) == ether_type_arp
+end
+
+local function ipv4_eq(a, b) return C.memcmp(a, b, 4) == 0 end
+local function mac_eq(a, b)  return C.memcmp(a, b, 6) == 0 end
```
Comment (dpino) [inline src/apps/ipv4/arp.lua]: `mac_eq` is never used.

**Comment by wingo** [comment]
I believe the CI error here is spurious:
```
ERROR snabbnfv-iperf-1500-tunnel+crypto -> 0.847917 of 0.96 (SD: 0.0363318 )
```
There's nothing that this patch touches that's related to that.

**Pull request closed without merging**